bot/cogs/utils/api.py

```diff
@@ -20,133 +38,154 @@ def __init__(self, id, ip, port, web_url=None):
+        """
 
-    @property
-    def headers(self):
-        """ Default authentication header the API needs. """
-        return {'authentication': self.api_key}
+        if Config.api_url:
+            return f'{Config.api_url}/match/{self.id}'
 
-    async def generate_link_url(self, user_id):
-        """ Get custom URL from API for user to link accounts. """
-        url = f'{self.base_url}/discord/generate/{user_id}'
 
-        async with self.session.get(url=url, headers=self.headers) as resp:
+class User:
+    def __init__(self, user_id: int) -> None:
+        """
+
+        Parameters
+        ----------
+        user_id : int
+        """
+
+        self.user_id = user_id
+
+    async def generate_link_url(self) -> str:
+        """Get custom URL from API for user to link accounts.
+
+        Returns
+        -------
+        str
+            Formatted link.
+        """
+
+        url = f'{Config.api_url}/discord/generate/{self.user_id}'
+
+        async with Sessions.requests.get(url=url) as resp:
             resp_json = await resp.json()
 
-            if resp_json.get('discord') and resp_json.get('code'):
-                return f'{self.base_url}/discord/{resp_json["discord"]}/{resp_json["code"]}'
+            if "discord" in resp_json and "code" in resp_json:
+                return f'{Config.api_url}/discord/{resp_json["discord"]}/{resp_json["code"]}'
+
+    async def is_linked(self) -> bool:
+        """
+        Returns
+        -------
+        bool
+        """
 
-    async def is_linked(self, user_id):
-        """ Check if a user has their account linked with the API. """
-        url = f'{self.base_url}/discord/check/{user_id}'
+        url = f'{Config.api_url}/discord/check/{self.user_id}'
 
-        async with self.session.get(url=url, headers=self.headers) as resp:
+        async with Sessions.requests.get(url=url) as resp:
             resp_json = await resp.json()
 
-        if resp_json.get('linked'):
-            return resp_json['linked']
-        else:
-            return False
+            return resp_json["linked"] if "linked" in resp_json else False
+
+    async def get_player(self) -> Player:
+        """Get player data from the API.
+
+        Returns
+        -------
+        Player
+        """
+
+        url = f'{Config.api_url}/player/discord/{self.user_id}'
+
+        async with Sessions.requests.get(url=url) as resp:
+            return Player(await resp.json())
+
+
+async def update_discord_name(user: Player) -> bool:
+    """Update a users API name to their current Discord display name.
+
+    Parameters
+    ----------
+    user : Player
+
+    Returns
+    -------
+    bool
+    """
+
+    url = f'{Config.api_url}/discord/update/{user.id}'
+    data = {"discord_name": user.display_name}
+
+    async with Sessions.requests.post(url=url, data=data) as resp:
+        return resp.status == 200
+
+
+async def get_players(user_ids: list) -> AsyncGenerator[Player, None]:
```

Comment:
I presume this will now let the return value be iterated over before all the results are returned? This should be very nice for scaling up.

Reply:
Yeah ideally `StatsCog.leaders` should be utilized to take advantage of this more.

Reply:
How can I take better advantage of the async generator? The player data is grabbed all at once so it's not like there is much waiting that can be done once we have it all.

Reply:
https://github.com/WardPearce/csgo-league-bot/blob/utils-refactor/bot/cogs/stats.py#L69

Instead of looping it into a list, you'd be better off doing something what used the generator the format the message.

e.g.
```py
async def leaders(self, ctx):
    async for player in get_players_stats([user.id for user in ctx.guild.members]):
        ...
```

Reply:
I don't think that would work because I need to sort them by score first to figure out the top X players to put on the leaderboard. After that I am only iterating over 5 players anyways.

Reply:
Yeah, but you could solve that issue by updating the API to return the players ordered by elo & also add a limit so you aren't processing a crap ton of unneeded information.

Reply:
Yeah I think having an endpoint to get the top X players sorted by rating would be helpful. For now this is the best the bot can do.